packages/material-ui/src/Divider/Divider.js

```diff
@@ -36,11 +36,23 @@ export const styles = theme => ({
 });
 
 const Divider = React.forwardRef(function Divider(props, ref) {
-  const { absolute, classes, className, component: Component, light, variant, ...other } = props;
+  const {
+    absolute,
+    classes,
+    className,
+    component: Component,
+    light,
+    tabIndex: tabIndexProp,
+    variant,
+    ...other
+  } = props;
 
   if (Component === 'li' && !other.role) {
     other.role = 'separator';
   }
+  // tabIndex defaults to -1 in the DOM. Using something lower to indicate to
+  // MenuList to skip it in keyboard focus navigation.
```

Comment:
If it should be skipped it shouldn't have any tabIndex: https://www.w3.org/TR/2009/WD-wai-aria-practices-20090224/#focus_tabindex. I didn't find any documentation that the -1 default comes from the DOM.

Reply:
> I didn't find any documentation that the -1 default comes from the DOM.

Neither did I -- I only discovered this when my tests didn't work and then I did some experiments in both React and plain html/JavaScript to verify. Here's one example I created when trying to understand the actual behavior: https://codesandbox.io/s/2jo7yxrr2n

Reply:
I just found the following [here](https://html.spec.whatwg.org/multipage/interaction.html#attr-tabindex):

> When the attribute is omitted, the user agent applies defaults. (There is no way to make an element that is being rendered be not focusable at all without disabling it or making it inert.)

Reply:
Right I always forget that there is a difference between DOM properties and HTML attributes: https://html.spec.whatwg.org/multipage/interaction.html#dom-tabindex defines `HTMLElement.tabIndex` to default to -1.

What we're looking for is `HTMLElement.getAttribute('tabIndex')`. I would prefer this over defining our own values.

Reply:
Sounds good. I'll rework it. I didn't like defining my own values either.

Reply:
@eps1lon I still have one sticking point around disabled menu items. `MenuItem` uses `<ListItem button/>` which leverages `ButtonBase`. `ButtonBase` always uses -1 tabIndex for disabled buttons. It can't leave tabIndex undefined, because for a `button` tabIndex would default to 0 and then be tabbable. The `disabled` property currently isn't propagated all the way to the component, and I suspect it needs to stay that way.

At the moment I have things working with the following in ButtonBase:
```
data-mui-disabled={disabled ? true : undefined}
```
and then in MenuList:

```
    const hasExplicitOrTabbableTabIndex = nextFocus.tabIndex >= 0 || nextFocus.hasAttribute("tabindex");
    if (!hasExplicitOrTabbableTabIndex || nextFocus.disabled || nextFocus.hasAttribute("data-mui-disabled")) {
      nextFocus = traversalFunction(list, nextFocus, disableListWrap);
    } else {
      break;
    }
```
This seems preferable to the -2 tabIndex, but not sure how much concern there is around adding a data attribute like this.

Reply:
Unfortunately, with changing to using attributes, I'm running into browser differences that are difficult to work with. As I mentioned above, the absence of a `tabindex` attribute is not sufficient due to elements that are inherently tabbable without a `tabindex` attribute where specifying a `tabindex` attribute less than zero is the only way to remove them from the tab order.

I thought I would still be able to get things to work using a more complex heuristic involving a combination of the `tabindex` attribute and the `tabIndex` DOM property being zero or greater, but Edge and IE have bizarre behavior around the tabIndex DOM property. Instead of defaulting to -1 like Chrome, they default to 0 even when the element isn't tabbable. In Edge (and I think IE as well), the tabIndex DOM property is zero for an `hr` element if the attribute is missing and also if I explicitly specify it as zero, but it is only tabbable in the latter case.

For now, I'm going to roll back to my -2 tabindex approach (because it worked reliably even if slightly hackish in nature), but I'm open to other ideas.

Reply:
@eps1lon You can ignore much of my last couple of comments. I was missing that a disabled `ButtonBase` propagates **either** the `disabled` attribute or the `aria-disabled` attribute. This gives me existing ways to reliably check for "disabled" and then in combination with the `tabindex` attribute existence, I was able to get all my tests working for dividers and disabled menu items without any -2 tabindex weirdness or new, redundant attributes.

Reply:
Nice. I think we shouldn't set the tabIndex to -1 on the ButtonBase (or any other element) if it's disabled. Maybe that's the better option than checking for tabIndex, disabled and aria-disabled. 

What I'm still missing is how we can acces the "focus flag" that the HTML spec mentions. Maybe that's an internal flag? But I'm ok with the current solution. If we need an additional helper attribute I'd rather prefer some `data-mui-*` attribute instead of hijacking `tabIndex`.

Reply:
I suspect the reason for the -1 is to ensure it is taken out of the focus order. That isn't necessary when the component type is 'button' (since being disabled should take it out of the focus order), but when the component type is not 'button' I don't think `aria-disabled` will remove it from the focus order if it is something inherently tabbable (e.g. 'a' tag), whereas setting the tabindex to -1 will reliably take it out of the focus order.

Reply:
Here's a quick demonstration: https://codesandbox.io/s/3v6wq38q71

Reply:
> What I'm still missing is how we can access the "focus flag" that the HTML spec mentions. 

Far as I can tell it is only internal. The tabIndex DOM property is the closest thing and seems reliable as the "focus flag" on Chrome, but not Edge or IE.